pysoa/client/client.py

```diff
@@ -4,22 +4,142 @@
+        """
+        Given a list of in-order middleware callables `middleware`
+        and a base function `base`, chains them together so each middleware is
+        fed the function below, and returns the top level ready to call.
+        """
+        for ware in reversed(middleware):
+            base = ware(base)
+        return base
+
+    def _base_send_request(self, request_id, meta, job_request):
+        if isinstance(job_request, JobRequest):
+            job_request = attr.asdict(job_request)
+        message = self.serializer.dict_to_blob(job_request)
+        self.transport.send_request_message(request_id, meta, message)
+
+    def send_request(self, job_request):
+        """
+        Serialize and send a JobRequest, and return a request ID.
+
+        The context and control_extra arguments may be used to include extra values in the
+        context and control headers, respectively.
+
+        Args:
+            job_request: JobRequest object
+        Returns:
+            int
+        Raises:
+            ConnectionError, InvalidField, MessageSendError, MessageSendTimeout,
+            MessageTooLarge
+        """
+        request_id = self.request_counter
+        self.request_counter += 1
+        meta = self._prepare_metadata()
+        wrapper = self._make_middleware_stack(
+            [m.request for m in self.middleware],
+            self._base_send_request,
+        )
+        wrapper(request_id, meta, job_request)
+        return request_id
+
+    def _get_response(self):
+        request_id, meta, message = self.transport.receive_response_message()
+        if message is None:
+            return (None, None)
+        else:
+            raw_response = self.serializer.blob_to_dict(message)
+            job_response = JobResponse(**raw_response)
+            return request_id, job_response
+
+    def get_all_responses(self):
+        """
+        Receive all available responses from the trasnport as a generator.
+
+        Yields:
+            (int, JobResponse)
+        Raises:
+            ConnectionError, MessageReceiveError, MessageReceiveTimeout, InvalidMessage,
+            StopIteration
+        """
+        wrapper = self._make_middleware_stack(
+            [m.response for m in self.middleware],
+            self._get_response,
+        )
+        while True:
+            request_id, response = wrapper()
+            if response is None:
+                break
+            yield request_id, response
+
+
 class Client(object):
     """The Client provides a simple interface for calling actions on Servers."""
 
-    def __init__(self, service_name, transport, serializer, middleware=None, context=None):
-        self.service_name = service_name
-        self.transport = transport
-        self.serializer = serializer
-        self.middleware = middleware or []
-        self.context = context
-        self.request_counter = 0
+    settings_class = ASGIClientSettings
+    handler_class = ServiceHandler
+
```

Comment:
A `config` is basically the routing information for each service, correct? Right now, it appears that each client is being initialized with `ASGIClientSettings` by default. Does this work for local services? The client now acts as the router, so you should be able to call any service for which transport and serialization information has been configured.

My original request for `config` schema validation was meant to catch errors in the `config` dictionary format (i.e. not `{<service name key>: <client config dict>, ...}`).